### DISCUSSION TITLE
ibus-mozc doesn't work when switching from Russian layout

### DISCUSSION THREAD

**Discussion by WGH-** [google/mozc#408] [Q&A] [answered]
I have three layouts in ibus: `['xkb:us::eng', 'mozc-jp', 'xkb:ru::rus']`.

mozc only works correctly when switching from English. If you switch from Russian layout to mozc, it will type Cyrillic letters instead of functioning properly.

```
% ibus version
IBus 1.5.14
```

It happens on at least mozc-2.16.2037.102 and mozc-2.20.2677.102.

**Comment by WGH-** [comment]
FWIW, anthy doesn't suffer from this.

**Comment by WGH-** [comment]
This is caused by layout option in `/usr/share/ibus/component/mozc.xml`. So basically if you change layout from `default` to `jp`, it works correctly. Anthy uses `jp` as well, which explains why Anthy doesn't have this issue.

```diff
diff --git a/component/mozc.xml b/-
index a7be498..0000000 100644
--- a/component/mozc.xml
+++ b/-
@@ -16,7 +16,7 @@
   <icon_prop_key>InputMode</icon_prop_key>
   <icon>/usr/share/ibus-mozc/product_icon.png</icon>
   <setup>/usr/lib64/mozc/mozc_tool --mode=config_dialog</setup>
-  <layout>default</layout>
+  <layout>jp</layout>
   <name>mozc-jp</name>
   <longname>Mozc</longname>
 </engine>
```

**Comment by WGH-** [comment]
Ping. Any reasons why `default` should be used instead of `jp`?

**Comment by WGH-** [comment]
Some related discussion can be found here: ibus/ibus#1943

**Comment by hiroyuki-komatsu** [comment, marked as answer]
Hi, sorry for too late reply,

The `default` basically  works well unless the user uses multiple keyboard layouts like this issue.
If we use `jp` here, people who use the `us` layout will have an inconsistent layout problem for example.

As you mentioned, you can specify the keyboard layout by modifying `mozc.xml` if you have the root permission.

Since 9ba59b64d53365c1fe93c1c245b4ec3e35bdadf0, the setting is moved to your home directory.
So you do not need the root permission.

**Comment by homocomputeris** [comment]
Hi, I have the same problem with an additional issue: I use Dvorak.

If I configure Mozc either as
```shell
$ cat $HOME/.config/mozc/ibus_config.textproto           
# `ibus write-cache; ibus restart` might be necessary to apply changes.
engines {
  name : "mozc-jp"
  longname : "Mozc"
  layout : "us"
  layout_variant : "dvorak"
}
active_on_launch: False
```
or
```shell
$ cat $HOME/.config/mozc/ibus_config.textproto           
# `ibus write-cache; ibus restart` might be necessary to apply changes.
engines {
  name : "mozc-jp"
  longname : "Mozc"
  layout : "jp"
  layout_variant : "dvorak"
}
active_on_launch: False
```
the input layout (before conversion to Japanese) is always QWERTY.

If I set
```shell
$ cat $HOME/.config/mozc/ibus_config.textproto           
# `ibus write-cache; ibus restart` might be necessary to apply changes.
engines {
  name : "mozc-jp"
  longname : "Mozc"
  layout : "default"
  layout_variant : ""
}
active_on_launch: False
```
then the input layout is the last used one: either Dvorak or Russian.

Any hint how to make it work with Dvorak?

**Reply by hiroyuki-komatsu** [reply]
Hi, would you check the output of `/usr/lib/ibus-mozc/ibus-engine-mozc --xml`, and check if the output contains the `<layout_variant>` tag?

Here's my output for example:
```
<engines>
<engine>
  <description>Mozc (Japanese Input Method)</description>
  <language>ja</language>
  <icon>/usr/share/ibus-mozc/product_icon.png</icon>
  <rank>80</rank>
  <icon_prop_key>InputMode</icon_prop_key>
  <symbol>&#x3042;</symbol>
  <setup>/usr/lib/mozc/mozc_tool --mode=config_dialog</setup>
  <name>mozc-jp</name>
  <longname>Mozc</longname>
  <layout>default</layout>
  <layout_variant></layout_variant>
  <layout_option></layout_option>
</engine>
</engines>
```

if your output doesn't contain `<layout_variant>`, your Mozc version doesn't support it yet.
In that case, you might want to do either of

1. Update the Mozc version.
2. Modify `/usr/share/ibus/component/mozc.xml` as a workaround.

The file locations are possibly different if your distribution is not Debian or Ubuntu.

**Reply by homocomputeris** [reply]
At the moment, I have the following settings:
`/usr/lib/mozc/mozc_tool --help` shows ver. 2.26.4538.102
Built on Arch Linux from the commit 5e7fdb7551b1928d09eaf2114f19601458bc6c31.

```
cat $HOME/.config/mozc/ibus_config.textproto

# `ibus write-cache; ibus restart` might be necessary to apply changes.
engines {
  name : "mozc-jp"
  longname : "Mozc"
  layout : "jp"
  layout_variant : "dvorak"
  rank : 0
}
active_on_launch: False
```

```
/usr/lib/ibus-mozc/ibus-engine-mozc --xml

<engines>
<engine>
  <description>Mozc (Japanese Input Method)</description>
  <language>ja</language>
  <icon>/usr/share/ibus-mozc/product_icon.png</icon>
  <rank>0</rank>
  <icon_prop_key>InputMode</icon_prop_key>
  <symbol>&#x3042;</symbol>
  <setup>/usr/lib/mozc/mozc_tool --mode=config_dialog</setup>
  <name>mozc-jp</name>
  <longname>Mozc</longname>
  <layout>jp</layout>
  <layout_variant>dvorak</layout_variant>
  <layout_option></layout_option>
</engine>
</engines>
```

Which somehow results in:

- Mozc Direct input: QWERTY instead of Dvorak.
- Mozc Hiragana: QWERTY instead of Dvorak.
